Add runner option to ament_add_test

pbaughman opened the pull request:
Allow the user of `ament_add_gtest` and `ament_add_pytest_test` to specify a different test runner.  This will be used by ROS2 tests so they can optionally specify a runner that provides isolation between tests running in parallel

(resolves #172)

### A note about changing dependencies:
Previously, the dependencies for ament_cmake_test looked like this:
```
$ colcon list --topological-graph --packages-up-to ament_cmake_test
ament_package     +*.                                                                                          
ament_cmake_core   +*                                                                                          
ament_cmake_test    +                                                                                          
```
In this PR, I changed run_test.py from a python script to a python script that imports most of its functionality from a new python library called `ament_add_test`.  This required me to change the dependencies of ament_cmake_test to depend on ament_cmake_python.  Now the deps look like this:

```
$ colcon list --topological-graph --packages-up-to ament_cmake_test
ament_package       +*..                                                                                          
ament_cmake_core     +**                                                                                          
ament_cmake_python    +*                                                                                          
ament_cmake_test       +                                                                                          
```
As you can see, this did not cause many other indirect dependencies.  Since ament_cmake_test relied on python anyway (to execute run_test.py) I think this is OK.

pbaughman commented:
@tfoote This is the first of two PRs required to implement isolated testing for ROS2.  This change is ROS agnostic, but will allow us to have wrappers around `ament_add_pytest_test` and `ament_add_gtest` that set up the test isolation

@wjwwood This is the approach we discussed last week that keeps ROS stuff out of ament.  I will create a follow-up PR in ament_cmake_ros with the aforementioned wrappers.

wjwwood commented:
@dirk-thomas can you comment on the change in the dependency structure due to this change? Is it ok or is there a better way to do this to avoid that dependency?

dirk-thomas commented on ament_cmake_test/ament_cmake_test/__init__.py:
```diff
@@ -0,0 +1,351 @@
+# Copyright 2014-2015 Open Source Robotics Foundation, Inc.
```
Currently the history of this file is lost. Please copy the content of the file without any changes in a separate commit. Then in a second commit make any necessary modifications to the original file as well as the new one. That should allow git to track the history of the content.

pbaughman replied:
@dirk-thomas I'm not seeing the history in still. . . I think maybe I have to do a 'move' operation instead of a copy.  Can you confirm that the history is still missing?

dirk-thomas replied:
Try looking at the blame view instead. The history view might not work correctly either way.

pbaughman replied:
Blame shows that I wrote 100% of the lines five hours ago.  I'll try a move operation followed by an edit so at least ament_cmake_test/ament_cmake_test/__init__.py will have history.  Since that's discoverable from run_test.py, I think that should be enough

dirk-thomas commented on ament_cmake_test/cmake/ament_add_test.cmake:
```diff
@@ -50,7 +50,7 @@
   cmake_parse_arguments(ARG
     "GENERATE_RESULT_FOR_RETURN_CODE_ZERO;SKIP_TEST"
-    "OUTPUT_FILE;RESULT_FILE;TIMEOUT;WORKING_DIRECTORY"
+    "OUTPUT_FILE;RESULT_FILE;RUNNER;TIMEOUT;WORKING_DIRECTORY"
```
Same here - docblock.

pbaughman commented on ament_cmake_gtest/cmake/ament_add_gtest_test.cmake:
```diff
@@ -83,6 +83,7 @@ function(ament_add_gtest_test target)
     COMMAND ${cmd}
     OUTPUT_FILE "${CMAKE_BINARY_DIR}/ament_cmake_gtest/${target}.txt"
     RESULT_FILE "${result_file}"
+    RUNNER "${ARG_RUNNER}"
```
I'll make it match the style of the other arguments. Note that `ament_add_gtest` does things differently than `ament_add_gtest_test` and `ament_add_pytest_test` even before my change

pbaughman commented:
@dirk-thomas Ok, I think the history is fixed now.  Git blame on github, and on my machine now shows you and William wrote most of the code of `ament_cmake_test/__init__.py` 4  years ago.

pbaughman commented:
@dirk-thomas Ok, I think this is ready to go.  I believe I've addressed all of your comments and I tested it along with https://github.com/ros2/ament_cmake_ros/pull/3 and verified that rcl and rclpy can use the new 'isolated' version of the test macros successfully

dirk-thomas commented on ament_cmake_gtest/cmake/ament_add_gtest.cmake:
```diff
@@ -25,6 +25,8 @@
 # :type target: string
 # :param ARGN: the list of source files
 # :type ARGN: list of strings
+# :param RUNNER: the path to a non default test runner to use
```
Since the user could pass the default test runner explicitly please update the line to:

> :param RUNNER: the path to the test runner script (default: see `ament_add_test`).

Same below.

dirk-thomas commented on ament_cmake_test/cmake/ament_add_test.cmake:
```diff
@@ -25,6 +25,8 @@
 # :type COMMAND: list of strings
 # :param OUTPUT_FILE: the path of the file to pipe the output to
 # :type OUTPUT_FILE: string
+# :param RUNNER: the path to a non default test runner to use (default uses run_test.py)
```
In this case use the actual default:

> :param RUNNER: the path to the test runner script (default: run_test.py).

pbaughman commented:
@dirk-thomas :+1:

dirk-thomas commented on ament_cmake_gtest/cmake/ament_add_gtest_test.cmake:
```diff
@@ -20,6 +20,8 @@
 #
 # :param target: the target name which will also be used as the test name
 # :type target: string
+# :param RUNNER: the path to a non default test runner to use
```
This line needs to be updated too.

dirk-thomas commented:
Since we don't want to squash on merge can you please squash the commit 4 (and soon 5) into the first commit. Thanks.

pbaughman commented:
@dirk-thomas Done

dirk-thomas commented:
Sanity check to rule out regressions: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7303)](https://ci.ros2.org/job/ci_linux/7303/)

dirk-thomas commented:
Thanks for the enhancement.
